### ISSUE TITLE
[discussion] add —watch to the wireit command as an option

### ISSUE THREAD

**Issue by michaelwarren1106** [google/wireit#553]
was curious about the reason behind the —watch flag not being able to be added to the wireit command directly?

imo, it would be super useful to be able to do

```
“scripts”: {
  “dev”: “wireit —watch” //always watch mode
}
```

imo it’s a bit clunky to have to send the “--watch” flag with the runner command , ie npm/yarn/pnpm run rather than being able to bake it into the command by default. 

i have a command similar to the above and i always forget that i need to call it with the —watch flag manually.

**Comment by Westbrook** [comment]
I've also noticed this and wondered if this should actually be part of the `wireit` config. It happens that almost all of my `--watch` commands are services that deserve a watch flag in their own "command" such that they can't really beat  run without `--watch` anyways...

**Comment by michaelwarren1106** [comment]
+1 to a `watch: true` config option as well. imo either could be great

**Comment by evelant** [comment]
A `watch: true` would be useful, when I first read the documentation I felt it was strange that this wasn't already possible

**Comment by augustjk** [comment]
I think these are valid suggestions. I suppose workaround for now would be to do something like
```json
"scripts": {
  "dev:base": "wireit",
  "dev": "npm run dev:base --watch"
}
```
so `npm run dev` will always watch.

**Comment by aomarks** [comment]
Yeah, I'm open to either `"watch": "wireit --watch"` or something like `{"watch": true}` which would be more convenient and also eliminate the extra node process needed by the workarounds.

My only trepidation is around using such a script as a *dependency*, though. Watch mode is a global thing -- it applies to all scripts in the graph. So what would it mean to launch a script *not* in watch mode, but *depend* on a script that is "intrinsically" in watch mode?

Maybe the simple answer to that is that we don't allow making a intrinsic-watch script a dependency, which would force you to factor out a separate e.g. `:watch` script for something that can be used in either way.

**Comment by evelant** [comment]
That sounds like a great solution.

**Comment by VanCoding** [comment]
@aomarks Maybe a bit off-topic: What if we would allow specifying a "watchCommand" next to the normal "command". The "watchCommand" would be expected to be long-running and would be executed instead of the normal command, when the "--watch" flag gets passed.

This way, it wouldn't be necessary to have two separate commands for watching. For normal builds it would be `npm run build` and for watching it would be `npm run build --watch`. It wouldn't directly solve the original problem, but it probably would make it less painful. But I haven't fully thought this throug, it probably would mean changing other stuff.

**Comment by filimon-danopoulos-stratsys** [comment]
Adding my voice to this, I have some trouble with my peers using the wrong commands and this would help tremendously.

**Comment by jonthenerd** [comment]
Just voicing support for the original proposal of having a package.json script call `wireit --watch` . This would ease transition for those developers coming from setups where starting to develop normal entails running `npm run dev`

The current workaround proposed doesn't work cleanly on Windows 11 + Windows Terminal + PowerShell Core, as it produces repetition and looks like something is stuck in a loop, so bad DevX.

**Comment by appsforartists** [comment]
I figured it was perhaps time to refactor
```json
"start": "yarn run -TB concurrently \"yarn run build:tsc --watch\" \"rollup --config ./rollup.config.js --watch\" \"dhost site --bind-all\"",
```
to use the now-available `service` mode.  However, I'm not sure how to wire it up.

I tried using @augustjk's workaround:

```
  "scripts": {
    "start": "yarn run serve --watch",
    "serve": "yarn run -TB wireit",
    "build": "yarn run -TB wireit",
    "build:tsc": "yarn run -TB wireit",
    "build:bundle": "yarn run -TB wireit",
    "publish": "yarn run -TB wireit"
  },
  "wireit": {
    "serve": {
      "service": true,
      // i've tried cascade: false here too
      "command": "dhost site --bind-all",
      "dependencies": [
        "build"
      ]
    },
    "build": {
      "dependencies": [
        "build:tsc",
        "build:bundle"
      ]
    },
```

however, it seems like `service: true` is blocking `watch` from doing its job:

```
🏃 [build:tsc] Running command "tsc --build --pretty"
✅ [build:tsc] Executed successfully
🏃 [build:bundle] Running command "rollup --config ./rollup.config.js"

./dist/mount.js → ./site/bundles...
created ./site/bundles in 69ms
✅ [build:bundle] Executed successfully
✅ [build] No command to execute
🏃 [serve] Running command "dhost site --bind-all"
* Serving static files from …
* Local http://localhost:9000 (on your clipboard!)
* Network http://192.168.86.193:9000
------------------------------------------
^C💀 [serve] Killed
👀 [serve] Watching for file changes
```

Notice that 👀  doesn't show up until after I press `^C` to stop `wireit`.  `serve` is just a dumb server, but I want the dependencies (`build`) to run again on file changes.

There are other issues that would need resolution (like making sure that `livereload` still works correctly), but I can't make any progress until I figure out how `--watch` and `service: true` are meant to work together.

Not sure if this is a bug, a misunderstanding, or an unimplemented feature.  Happy to raise into its own issue if that's a better venue.

**Comment by jsudelko** [comment]
> So what would it mean to launch a script not in watch mode, but depend on a script that is "intrinsically" in watch mode?
Maybe the simple answer to that is that we don't allow making an intrinsic-watch script a dependency, which would force you to factor out a separate e.g. :watch script for something that can be used in either way.

As others in the thread have hinted, it's common to have a service that is not watched by Wireit that is dependent on packages that should be watched by Wireit (which might in turn depend on other services). Some packages in the graph may need to be watched/managed by other tools such Vite/Rollup (and not Wireit) so that they can handle incremental bundling, HMR etc.

A simple work around might be to just provide an empty `files`/`output` array in the Wireit config depending on your needs, so that it's not fighting with your bundler etc. over outputs and inputs.
